kong/core/router.lua

```diff
@@ -619,30 +621,18 @@ function _M.new(apis)
     end
 
 
-    upstream_host = api_t.upstream_host
-
-
     if api_t.strip_uri_regex then
       local stripped_uri = re_sub(uri, api_t.strip_uri_regex, "/$1", "jo")
       ngx.req.set_uri(stripped_uri)
     end
 
 
-    if api_t.preserve_host then
-      if not headers then
-        headers = ngx.req.get_headers()
-      end
-
-      upstream_host = headers["host"]
-    end
-
-
     if ngx.var.http_kong_debug then
       ngx.header["Kong-Api-Name"] = api_t.api.name
     end
 
 
-    return api_t.api, api_t.upstream_scheme, upstream_host, api_t.upstream_port
+    return api_t.api, api_t.upstream_scheme, api_t.upstream_host, api_t.upstream_port
```

Comment:
At the time of writing, the goal of this API was to return everything needed in order for Kong to forward upstream. Now with the balancing bit in between and the removal of the `preserve_host` logic from this module, it seems like this is un-necessarily bloated. 

We should either continue encapsulating such logic in this module (might not be possible since we now need to the port from the DNS resolution to build the definitive upstream Host header), **or**, simply return the Kong API (`api_t.api`) and `api_t` itself (which has all the necessary properties, but exposes some underlying router implementation which we don't really want).

We should return a table containing the final `api` + eventually another table containing `scheme`, `host` and `port`, **or** include those properties in the 1st table being returned. Those are minor modifications.